sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/ReactorNettyHttpResponseBase.java

```diff
@@ -31,8 +42,136 @@ public final String getHeaderValue(String name) {
 
     @Override
     public final HttpHeaders getHeaders() {
-        HttpHeaders headers = new HttpHeaders();
-        reactorNettyResponse.responseHeaders().forEach(e -> headers.put(e.getKey(), e.getValue()));
+        if (headers == null) {
+            headers = new NettyHttpHeaders(reactorNettyResponse.responseHeaders());
+        }
         return headers;
     }
+
+    static class NettyHttpHeaders extends HttpHeaders {
+        private final io.netty.handler.codec.http.HttpHeaders nettyHeaders;
+        private Map<String, String> map;
+
+        NettyHttpHeaders(io.netty.handler.codec.http.HttpHeaders nettyHeaders) {
+            this.nettyHeaders = nettyHeaders;
+        }
+
+        private String formatKey(final String key) {
+            return key == null ? null : key.toLowerCase(Locale.ROOT);
+        }
+
+        @Override
+        public int getSize() {
+            return nettyHeaders.size();
+        }
+
+        /* This replaces any current value with the given name */
+        @Override
+        public HttpHeaders put(String name, String value) {
+            if (name == null) {
+                return this;
+            }
+
+            name = formatKey(name);
+            if (value == null) {
+                remove(name);
+            } else {
+                nettyHeaders.set(name, value);
+            }
+            return this;
+        }
+
+        HttpHeaders add(String name, String value) {
+            if (name == null) {
+                return this;
+            }
+
+            name = formatKey(name);
+            if (value == null) {
+                remove(name);
+            } else {
+                nettyHeaders.add(formatKey(name), value);
+            }
+            return this;
+        }
+
+        @Override
+        public HttpHeader get(String name) {
+            name = formatKey(name);
+            if (nettyHeaders.contains(name)) {
+                return new NettyHttpHeader(this, name, nettyHeaders.get(name));
```

Comment:
Are you proposing a new map to contain this?

Reply:
Yeah, lazily populate a map each time `get()` is called with a new `name`.

Reply:
I considered this when I was working on it. I also considered lazily populating a standard azure-core HttpHeaders implementation and then fully populating it in some of the later API (e.g. iterator, stream, and toMap). In the end I chose not to do the later, but I am ok with doing the former.

Reply:
Actually, thinking a little more, I think that this will lead to potential issues due to the fact that the netty headers may be modified independently from the cached HttpHeader. I'll leave this as a future optimisation.